### PR TITLE
Implement CacheConfig field permitting some operations to return non-expired cache data

### PR DESCRIPTION
**Pull request by dannycjones**
## Description of change

This plumbs in checks for if the filesystem should check with S3 to ensure we're consistent with S3 at open, create, etc.

There is no way to configure mountpoint-s3 itself to relax the consistency model - this change only impacts internals.

Relevant issues: #255

## Does this change impact existing behavior?

If the new field is set to false, it does impact the consistency model of the `mountpoint-s3::fs::S3Filesystem`. This is not configurable during mounting at this time, but may be in the future.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3/src/inode.rs [hunk 1]
```diff
@@ -1197,11 +1262,10 @@ impl From<InodeKind> for FileType {
...
     File {},
     Directory {
-        /// Mapping from child names to [Inode]s.
+        /// Mapping from child names to previously seen [Inode]s.
         ///
-        /// How should this field be used?:
-        /// - **Many operations should maintain** this list.
-        /// - **Only `mknod` and `mkdir` should read** this list, for checking if a file already exists.
+        /// The existence of a child or lack thereof does not imply the object does not exist,
+        /// nor that it current exists in S3 in that state.
```
Comment (passaro) [inline mountpoint-s3/src/inode.rs:1268]:
```suggestion
        /// nor that it currently exists in S3 in that state.
```

**Review by passaro — COMMENTED** [review]
Looks good. Mostly minor comments here and there.

Only slight concern is that we need to check the new cache config flag in quite a few different places. It seems to me that `lookup_by_name` could almost be the central place where we check it, if not for one case (when called by `getattr` on `open`).  
Not critical, but it could interesting to explore alternatives where the flag is only used in a single layer: e.g.  move the `getattr` code to a new `SuperblockInner::lookup_by_ino` method, which would also check the flag (to ignore `force_revalidate`) and then try the remote lookup.

File: mountpoint-s3/src/fs.rs [hunk 2]
```diff
@@ -194,6 +194,13 @@ impl<Client: ObjectClient> UploadState<Client> {
 
 #[derive(Debug, Clone)]
 pub struct CacheConfig {
+    /// Should the file system check S3 even when a valid cached entry may be available?
+    ///
+    /// When enabled, some operations such as `lookup` are allowed to be served from cache
```
[the remark this answers is not in the document]
Comment (dannycjones) [reply mountpoint-s3/src/fs.rs:199]: Ah, yeah!

File: mountpoint-s3/src/inode.rs [hunk 3]
```diff
@@ -518,14 +544,50 @@ impl SuperblockInner {
...
 
-        // TODO use caches. if we already know about this name, we just need to revalidate the stat
-        // cache and then read it.
-        let remote = self.remote_lookup(client, parent_ino, name).await?;
-        let lookup = self.update_from_remote(parent_ino, name, remote)?;
+        let lookup = if skip_cache {
+            None
+        } else {
+            let lookup = self.cache_lookup(parent_ino, name);
+            trace!("lookup returned from cache: {:?}", lookup);
```
[the remark this answers is not in the document]
Comment (dannycjones) [reply mountpoint-s3/src/inode.rs:551]: Sounds good. I can provide different trace message for each case.

File: mountpoint-s3/src/inode.rs [hunk 4]
```diff
@@ -518,14 +545,60 @@ impl SuperblockInner {
...
+        };
+
         lookup.inode.verify_child(parent_ino, name.as_ref())?;
         Ok(lookup)
     }
 
+    /// Lookup an [Inode] against known directory entries in the parent,
+    /// verifying any returned entry has not expired.
+    fn cache_lookup(&self, parent_ino: InodeNo, name: &str) -> Option<LookedUp> {
+        fn do_cache_lookup(parent: Inode, name: &str) -> Option<LookedUp> {
```
Comment (passaro) [inline mountpoint-s3/src/inode.rs:569]: nit: I'm not sure the local fn is appropriate here. Is there no way to make this logic more linear?

Comment (dannycjones) [reply mountpoint-s3/src/inode.rs:569]: I'm wanting to make the core logic clear while still letting us wrap it in traces and metrics.

The difficulty is that there's a lot of `Option` and `Result` unpacking, and so it gets pretty hard to follow if trying to eliminate the use of the `?` operator.

I gave it a go and played around for a bit, but I'm not happy with it:

```rust


    /// Lookup an [Inode] against known directory entries in the parent,
    /// verifying any returned entry has not expired.
    fn cache_lookup(&self, parent_ino: InodeNo, name: &str) -> Option<LookedUp> {
        let parent = self.get(parent_ino).ok()?;
        let lookup = match &parent.get_inode_state().ok()?.kind_data {
            InodeKindData::File { .. } => unreachable!("parent should be a directory!"),
            InodeKindData::Directory { children, .. } => {
                if let Some(inode) = children.get(name) {
                    let valid_stat = inode.get_inode_state().ok()
                        .map(|state| state.stat)
                        .filter(|stat| stat.is_valid());
                    if let Some(valid_stat) = valid_stat {
                        let lookup = LookedUp {
                            inode: inode.clone(),
                            stat: valid_stat.clone(),
                        };
                        Some(lookup)
                    } else {
                        None
                    }
                } else {
                    None
                }
            }
        };

        match &lookup {
            Some(lookup) => trace!("lookup returned from cache: {:?}", lookup),
            None => trace!("no lookup available from cache"),
        }
        metrics::counter!("metadata_cache.cache_hit", lookup.is_some().into());

        lookup
    }
```

I'll leave it as is unless you have any specific suggestions.

Comment (dannycjones) [reply mountpoint-s3/src/inode.rs:569]: Ah, I want `0` on cache miss for that metric. This way, you have from this counter both the number of hits (value of the counter) as well as the number of times we checked the cache (number of times the metric was emitted).

That's the tricky one to plumb in, otherwise your approach is the better option.
